starfish/experiment/builder/cli.py

```diff
@@ -21,54 +23,33 @@ def __call__(self, spec_json):
-    def add_to_parser(parser):
-        parser.add_argument(
-            "output_dir",
-            type=FsExistsType())
-        parser.add_argument(
-            "--fov-count",
-            type=int,
-            required=True,
-            help="Number of FOVs in this experiment.")
-        parser.add_argument(
-            "--hybridization-dimensions",
-            type=StarfishIndex(),
-            required=True,
-            help="Dimensions for the hybridization images.  Should be a json dict, with {}, {}, "
-                 "and {} as the possible keys.  The value should be the shape along that "
-                 "dimension.  If a key is not present, the value is assumed to be 0.".format(
-                Indices.ROUND.value,
-                Indices.CH.value,
-                Indices.Z.value))
-        for aux_image_name in AUX_IMAGE_NAMES:
-            arg = parser.add_argument(
-                "--{}-dimensions".format(aux_image_name),
-                type=StarfishIndex(),
-                help="Dimensions for the {} images.  Should be a json dict, with {}, {}, and {} as "
-                     "the possible keys.  The value should be the shape along that dimension.  If "
-                     "a key is not present, the value is assumed to be 0.".format(
-                    aux_image_name, Indices.ROUND, Indices.CH, Indices.Z))
-            Cli.name_arg_map[aux_image_name] = arg.dest
-        parser.set_defaults(starfish_command=Cli.run)
-
-        Cli.parser_group = parser
-
-    @staticmethod
-    def run(args, print_help=False):
-        write_experiment_json(
-            args.output_dir, args.fov_count, args.hybridization_dimensions,
-            {
-                aux_image_name: getattr(args, Cli.name_arg_map[aux_image_name])
-                for aux_image_name in AUX_IMAGE_NAMES
-            }
-        )
+def dimensions_option(name, required):
+    return click.option(
+        "--{}-dimensions".format(name),
+        type=StarfishIndex(), required=required,
+        help="Dimensions for the {} images.  Should be a json dict, with {}, {}, "
+             "and {} as the possible keys.  The value should be the shape along that "
+             "dimension.  If a key is not present, the value is assumed to be 0."
+             .format(
+             name,
+             Indices.ROUND.value,
+             Indices.CH.value,
+             Indices.Z.value))
+
+decorators = [
+    click.command(),
+    click.argument("output_dir", type=click.Path(exists=True, file_okay=False, writable=True)),
+    click.option("--fov-count", type=int, required=True, help="Number of FOVs in this experiment."),
+    dimensions_option("hybridization", True),
+]
+for image_name in AUX_IMAGE_NAMES:
+    decorators.append(dimensions_option(image_name, False))
+
+def build(output_dir, fov_count, hybridization_dimensions, **kwargs):
+    write_experiment_json(
+        output_dir, fov_count, hybridization_dimensions,
+        kwargs  # FIXME
```

Comment:
Ok, so our testing strategy would be to try to use the tool and make sure it works?

Reply:
I looked at the CLI explicitly and fixed previous versions to make the tests pass. I'm pretty happy to push away the `FIXME`, but having @ttung's eyes on this just in case I missed something was worth leaving it. 🙂